Implement genetic algorithm for gambler

Nikoleta-v3 opened the pull request:
Implement genetic algorithm for gambler.

This includes adding two methods to the archetype. The `crossover` method and the `mutation`. 

The crossover is quite straight forward. A random point is selected. The offspring takes the genes of the first parent to that point and the rest from the second parent.

For the mutation, there is a probability (`mutation_probability`) that each gene will change.
This is according to [line 73](https://github.com/Nikoleta-v3/axelrod-dojo/blob/1de0d45ae88deb325a5a883070894e55dbd95a6e/src/axelrod_dojo/archetypes/gambler.py#L73). Note that during the mutation a gene might exceed the value of 1 that s why I have a check in the next  line.

Tests for the methods and an integration tests for the algorithm with gambler have been implemented.
Feel free to squash and merge.

Commit:
Implement GA for Gambler archetype

Commit:
Implement GA for Gambler archetype

Commit:
Merge branch 'master' of github.com:Axelrod-Python/axelrod-dojo into ga-for-gambler

Commit:
Merge branch 'master' of github.com:Axelrod-Python/axelrod-dojo into ga-for-gambler

Commit:
tweak mutation

Commit:
tweak mutation

Commit:
add tests for gambler's ga methods

Commit:
add tests for gambler's ga methods

Commit:
Merge branch 'ga-for-gambler' of github.com:Nikoleta-v3/axelrod-dojo into ga-for-gambler

Commit:
fix mutation and adjust tests

Commit:
fix mutation and adjust tests

Commit:
add check to keep genes in bounds 0, 1

Commit:
add check to keep genes in bounds 0, 1

Commit:
add integration test

Commit:
add integration test

Commit:
tweak old tests

Commit:
Merge branch 'ga-for-gambler' of github.com:Nikoleta-v3/axelrod-dojo into ga-for-gambler

drvinceknight commented on src/axelrod_dojo/archetypes/gambler.py:
```diff
@@ -55,6 +66,27 @@ def create_vector_bounds(self):
 
+    @staticmethod
+    def mutate_pattern(pattern, mutation_probability):
+        for i in range(len(pattern)):
```
Better to use enumerate then `range(len`

```
for i, gene in enumerate(pattern):
    if random.random() <= mutation_probability:
        pattern[i] += (2 * random.random() - 1) 
```

Note that you are currently not quite using a uniform distribution (because of the `min(pattern[i], 1)`, I'd suggest:

```
for i, gene in enumerate(pattern):
    if random.random() <= mutation_probability:
        r = (2 * random.random() - 1) 
        if r >= 0:
            pattern[i] += r * (1 -  gene)
        else:
            pattern += r * gene
```

drvinceknight replied:
(This is because essentially if we're going to subtract `gene` is an upper bound and if we're going to add then `1 - gene` is an upper bound.)

Nikoleta-v3 replied:
This sounds good.

drvinceknight replied:
Actually @Nikoleta-v3 apologies I believe that is way more complicated than necessary, it's just a linear map from `[0, 1]` to `[-gene, 1-gene]` so:

```
for i, gene in enumerate(pattern):
    if random.random() <= mutation_probability:
        pattern[i] += random.random() - gene
```

Will do it:

- when `random.random()` is 1 we get `1 - gene`
- when `random.random()` is 0 we get `-gene`

Nikoleta-v3 replied:
Then this is equivalent to saying that:

```
pattern[i] = random.radnom()
```
I'll go with that. Thank you!

drvinceknight replied:
You're completely right! Apologies for making all that waaaay more complicated than it needed to be :+1:

drvinceknight commented on tests/algorithms/test_particle_swarm_optimization.py:
```diff
@@ -77,14 +77,11 @@ def test_pso_with_gambler(self):
-                                                          0.43758721,
-                                                          0.891773])))
-        self.assertEqual(abs(opt_objective_value), 3)
+        self.assertTrue(np.allclose(opt_vector, np.array([[0.72231352, 0.29956341,
```
What's happening here @Nikoleta-v3 ?

Nikoleta-v3 replied:
This is actually on me from my previous `pr`. My method `player` was not implemented correctly because I was not passing the parameters (`op_plays` etc ) just the pattern/vector. 

But I just realised that I am still not getting it quite right. Let me fix this and add specific tests to reassure that the `player` method works fine for both algorithms.

Commit:
test that player instance is created correctly

Commit:
test instance with vector

Tests that an instance of the Gambler is created correctly
when a vector is passed. That's when gambler works with
pso

Commit:
fix test on pso

This this the final tweak. Now the tests is correct because we
playing against 5 cooperators and the payoff we expect is around
5.

This was breaking previously because the player instance was not
created properly.

Commit:
typo

Commit:
fix mutation based on pr discussion.

Change the mutation to be:

- each gene has the probability to change to a number in [0, 1]

Remove the test about the genes beging between the bounds.

drvinceknight commented on src/axelrod_dojo/archetypes/gambler.py:
```diff
@@ -68,10 +68,9 @@ def create_vector_bounds(self):
-            if random.random() < mutation_probability:
-                pattern[i] += (2 * random.random() - 1) * pattern[i]
-                pattern[i] = min(pattern[i], 1)
+        for i, gene in enumerate(pattern):
```
You can go with `for i, _ in enumerate(pattern)` (as you don't use `gene`)

drvinceknight reviewed (approved):
Just the one stylistic suggestion from me @Nikoleta-v3 but looks good :+1:

Nikoleta-v3 commented:
Okay. Now the player instances (`Gambler`) are created correctly. For both the `pso` and the `ga`. I added specific tests in `gambler.py` to test the instances.

Commit:
fix style suggestion

marcharper commented on src/axelrod_dojo/archetypes/gambler.py:
```diff
@@ -38,13 +43,19 @@ def __init__(self, plays, op_plays, op_start_plays, pattern=None):
         """Receives a vector and creates an instance attribute called
         vector.  Ignores extra parameters."""
-        self.vector = vector
+        self.pattern = vector
```
Can we change the parameter / doc string to be consistent with the use of pattern instead of parameter? Or not change the attribute name.

marcharper commented on src/axelrod_dojo/archetypes/gambler.py:
```diff
@@ -55,6 +66,26 @@ def create_vector_bounds(self):
+    def mutate_pattern(pattern, mutation_probability):
+        for i, _ in enumerate(pattern):
+            if random.random() <= mutation_probability:
+                pattern[i] = random.random()
```
This is ok but it would work better if the mutation makes a minor adjustment rather than completely overwriting the value. In other words we're not trying to randomly jump around the space, rather we're exploring under the assumption that parameters close by each other typically have a similar fitness ("Hill-climbing").

Specifically, I would take the existing value of `pattern[i]`, add or subtract a random proportion between say -10% and 10%, and then normalize/truncate the value to [0,1] so that it's still a probability. The HMM player [does something like that](https://github.com/Axelrod-Python/axelrod-dojo/blob/master/src/axelrod_dojo/archetypes/hmm.py#L42).

For bonus points, you could write a function that could be used by both player types.

drvinceknight replied:
> This is ok but it would work better if the mutation makes a minor adjustment rather than completely overwriting the value. In other words we're not trying to randomly jump around the space, rather we're exploring under the assumption that parameters close by each other typically have a similar fitness ("Hill-climbing").

I suspect this is one of the things where we could go back and forth for a while. I would suggest that for consistency with the status quo of the dojo that you go with @marcharper's suggestion @Nikoleta-v3.

At some point we could probably rigorously study the parametrization of the various heuristics and see what works best :+1:) I might do this with my final year project student next year.

marcharper commented on src/axelrod_dojo/archetypes/gambler.py:
```diff
@@ -55,6 +66,26 @@ def create_vector_bounds(self):
+        pattern1 = self.pattern
+        pattern2 = other.pattern
+
+        midpoint = int(random.randint(0, len(pattern1)) / 2)
```
Why not `cross_point = random.randint(0, len(pattern1))` or the procedure that FSM uses [here](https://github.com/Axelrod-Python/axelrod-dojo/blob/master/src/axelrod_dojo/archetypes/fsm.py#L93)?

I assume the patterns are the same length, otherwise there are edge cases where you could choose beyond the end of `pattern2`. If so let's add a comment that states the assumption, or raise an exception if the assumption doesn't hold.

marcharper commented on src/axelrod_dojo/archetypes/gambler.py:
```diff
@@ -55,6 +66,26 @@ def create_vector_bounds(self):
+        pattern2 = other.pattern
+
+        midpoint = int(random.randint(0, len(pattern1)) / 2)
+        offspring = pattern1[:midpoint] + pattern2[midpoint:]
```
`offspring_pattern` (and below)

marcharper commented on tests/algorithms/test_particle_swarm_optimization.py:
```diff
@@ -71,20 +71,16 @@ def test_pso_with_gambler(self):
-                                                          0.43758721,
-                                                          0.891773])))
-        self.assertEqual(abs(opt_objective_value), 3)
+        self.assertTrue(np.allclose(opt_vector, np.array([[0. , 0. , 0.36158016,
```
Nice, I didn't know about `np.allclose`

marcharper commented on tests/archetypes/test_gambler.py:
```diff
@@ -21,6 +22,24 @@ def test_init(self):
+                                op_start_plays=op_start_plays, pattern=pattern)
+
+        player =  gambler.player()
+        action_dict = {Plays(self_plays=(C,), op_plays=(C,), op_openings=(C,)): 0,
```
Minor: could be cleaner / more succinct with a loop or generator clause.

marcharper commented on tests/integration/test_gambler.py:
```diff
@@ -0,0 +1,74 @@
+            for row in reader:
+                results.append(row)
+
+        self.assertTrue(all([0<=float(result[3]) <= 5 for result in results]))
```
nit: newline

marcharper commented on tests/integration/test_gambler.py:
```diff
@@ -0,0 +1,74 @@
+                                     output_filename=temporary_file.name,
+                                     opponents=self.opponents,
+                                     bottleneck=2,
+                                     mutation_probability = .01,
```
nit: spacing around `=`

marcharper commented on tests/archetypes/test_gambler.py:
```diff
@@ -100,3 +130,46 @@ def test_repr(self):
                          '1:1:1:0.05|0.07|0.7|0.03|0.6|0.7|0.45|0.1')
+
+    def test_crossover(self):
+        plays=1
```
nit: spacing around `=`

I like what you did above with `plays, op_plays, op_start_plays = 1, 1, 1`

marcharper commented on tests/archetypes/test_gambler.py:
```diff
@@ -100,3 +130,46 @@ def test_repr(self):
+                                       mutation_probability=0.8)
+
+        for _ in range(100):
+            _ = gambler_params.mutate()
```
I didn't know that this was a thing :+1:

Nikoleta-v3 commented:
Thank you for your comments. I will pick this up asap!

Commit:
address minor comments on pr

Commit:
fix crossover as suggested on pr

Commit:
fix mutation as suggested on pr

Commit:
add line at the end of file + pep8

Nikoleta-v3 commented:
@marcharper  I believe I addressed everything.

marcharper commented:
I still think the usage of vector versus pattern is potentially confusing at various places but it's not a big deal IMO. Thanks for this PR!
